controller/src/main/java/io/pravega/controller/store/stream/records/TagRecord.java

```diff
@@ -0,0 +1,135 @@
+
+@Data
+@Builder(toBuilder = true)
+public class TagRecord {
+
+    private static final TagRecordSerializer SERIALIZER = new TagRecordSerializer();
+    private final String tagName;
+    @Singular
+    private final Set<String> streams;
+
+    public static class TagRecordBuilder implements ObjectBuilder<TagRecord> {
+        public TagRecordBuilder removeStream(String stream) {
+            this.streams.remove(stream);
+            return this;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    public static TagRecord fromBytes(final byte[] data) {
+        return SERIALIZER.deserialize(data);
+    }
+
+    @SneakyThrows(IOException.class)
+    public byte[] toBytes() {
+        byte[] bytes = SERIALIZER.serialize(this).getCopy();
+        assert bytes.length < TableSegment.MAXIMUM_VALUE_LENGTH : "TagRecord is greater than the TableSegment Maximum value length";
+        return SERIALIZER.serialize(this).getCopy();
+    }
+    
+    private static class TagRecordSerializer
+            extends VersionedSerializer.WithBuilder<TagRecord,
+            TagRecord.TagRecordBuilder> {
+        @Override
+        protected byte getWriteVersion() {
+            return 0;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(0).revision(0, this::write00, this::read00);
+        }
+
+        @Override
+        protected void beforeSerialization(TagRecord streamConfigurationRecord) {
+            Preconditions.checkNotNull(streamConfigurationRecord);
+            Preconditions.checkNotNull(streamConfigurationRecord.getTagName());
+        }
+
+        private void read00(RevisionDataInput revisionDataInput, TagRecord.TagRecordBuilder configurationRecordBuilder)
+                throws IOException {
+            configurationRecordBuilder.tagName(revisionDataInput.readUTF());
+            byte[] comp = revisionDataInput.readArray();
+            Set<String> set = decompressArrayOption(comp);
+            configurationRecordBuilder.streams(set);
+        }
+
+        private void write00(TagRecord streamConfigurationRecord, RevisionDataOutput revisionDataOutput)
+                throws IOException {
+            revisionDataOutput.writeUTF(streamConfigurationRecord.getTagName());
+            byte[] comp = compressArrayOption(streamConfigurationRecord.getStreams());
+            revisionDataOutput.writeArray(comp);
+        }
+
+        @Override
+        protected TagRecord.TagRecordBuilder newBuilder() {
+            return TagRecord.builder();
+        }
+    }
+
+    private static byte[] compressArrayOption(final Set<String> set) throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        OutputStream out = new DeflaterOutputStream(baos);
+        for (String s : set) {
+            out.write(s.getBytes(StandardCharsets.UTF_8));
+            out.write(',');
+        }
+        out.flush();
+        out.close();
+        return baos.toByteArray();
```

Comment:
Tip:
Use `ByteBufferOutputStream`, then get the contents as a `ByteArraySegment`, then pass that to `revisionDataOutput.writeBuffer`. 

That will save you the trouble of all the extra array copies that happen when you invoke `toByteArray`. 

`writeBuffer/writeArray` are equivalent and both can be read back with `readArray` (like you already have)

Reply:
Thanks 👍  
Improved the code.